File: Gayeong/src/week4/BaekJoon7576.java
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

public class 토마토 {
    static int n, m;
    static int[][] graph;
    static Queue<Node> queue = new LinkedList<>();

    static int[] dx = {1, -1, 0, 0};
    static int[] dy = {0, 0, 1, -1};

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        m = sc.nextInt();
        sc.nextLine();
        graph = new int[m][n];

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                graph[i][j] = sc.nextInt();

                if (graph[i][j] == 1) queue.offer(new Node(i, j));
            }
            sc.nextLine();
        }

        bfs();

        int result = 0;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (graph[i][j] == 0) {
                    System.out.println(-1);
                    return;
                } else {
                    result = Math.max(result, graph[i][j]);
                }
            }
        }

        System.out.println(result - 1);
    }

    static class Node {
        int x;
        int y;

        public Node(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }
    static void bfs() {
        while (!queue.isEmpty()) {
            Node cur = queue.poll();

            for (int i = 0; i < 4; i++) {
                int nx = cur.x + dx[i];
                int ny = cur.y + dy[i];

                if (nx < 0 || nx >= m || ny < 0 || ny >= n || graph[nx][ny] != 0) continue;

                queue.offer(new Node(nx, ny));
                graph[nx][ny] = graph[cur.x][cur.y] + 1;
            }
        }

    }
}

File: Heepyo/src/week4/BaekJoon7576.kt
package week4

import java.util.*
import kotlin.math.max

private val startList: MutableList<IntArray> = ArrayList()

fun main() {
    val (m, n) = readln().split(" ").map(String::toInt)
    val array = Array(n + 2) { i ->
        if (i in (1..n))
            "-1 ${readln()} -1"
                .split(" ")
                .map(String::toInt)
                .also {
                    if (it.contains(1)) it.indices.forEach { j ->
                        if (it[j] == 1) startList.add(intArrayOf(i, j))
                    }
                }
                .toIntArray()
        else
            IntArray(m + 2) { -1 }
    }
    var answer = 0

    bfs(array)
    for (i in 1 until array.size - 1) {
        Arrays.sort(array[i])
        answer = if (array[i].binarySearch(0) > -1) {
            println(-1)
            return
        } else {
            max(answer, array[i][array[i].size - 1])
        }
    }
    println(answer - 1)
}

private fun bfs(array: Array<IntArray>) {
    val queue: Queue<IntArray> = LinkedList(startList)

    while (!queue.isEmpty()) {
        val outed = queue.poll()

        for (next in arrayOf(
            intArrayOf(outed[0] - 1, outed[1]),
            intArrayOf(outed[0], outed[1] + 1),
            intArrayOf(outed[0] + 1, outed[1]),
            intArrayOf(outed[0], outed[1] - 1)
        )) {
            if (array[next[0]][next[1]] == 0) {
                array[next[0]][next[1]] = array[outed[0]][outed[1]] + 1

                queue.offer(intArrayOf(next[0], next[1]))
            }
        }
    }
}